// ng-zorro-antd/esm2015/notification/nz-notification.definitions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function NzNotificationData() { }
function NzNotificationData_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    NzNotificationData.prototype.template;
    /** @type {?|undefined} */
    NzNotificationData.prototype.type;
    /** @type {?|undefined} */
    NzNotificationData.prototype.title;
}
/**
 * @record
 */
export function NzNotificationDataOptions() { }
function NzNotificationDataOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    NzNotificationDataOptions.prototype.nzKey;
    /** @type {?|undefined} */
    NzNotificationDataOptions.prototype.nzStyle;
    /** @type {?|undefined} */
    NzNotificationDataOptions.prototype.nzClass;
}
/**
 * @record
 */
export function NzNotificationDataFilled() { }
function NzNotificationDataFilled_tsickle_Closure_declarations() {
    /** @type {?} */
    NzNotificationDataFilled.prototype.messageId;
    /** @type {?|undefined} */
    NzNotificationDataFilled.prototype.state;
    /** @type {?|undefined} */
    NzNotificationDataFilled.prototype.options;
    /** @type {?} */
    NzNotificationDataFilled.prototype.createdAt;
}
